### PR TITLE
Link solidity libraries

### PR DESCRIPTION
**Pull request by jcarpanelli**
This is a first approach for linking solidity libs before deploying logic contracts.
Implemented so far:

- [x] Solidity libs discovery
- [x] Solidity libs linking
- [x] Solidity libs tracking (in `zos.network.json`)
- [x] Fixed `zos status ... --fetch` command to compare implementations as both logic contracts and solidity libs

Still have to:
- [ ] Add tests
- [ ] Fix skipped tests
- [ ] Finish fixing the `zos status ... --fix` command  
- [ ] Recursive solidity libs discovery

### CODE REVIEW

**Comment by spalladino** [comment]
@jcarpanelli please review the new commits!

File: packages/cli/test/mocks/mock-stdlib/build/contracts/GreeterLib.json [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+    "src": "0:130:1"
+  },
+  "compiler": {
+    "name": "solc",
+    "version": "0.4.24+commit.e67f0147.Emscripten.clang"
+  },
+  "networks": {},
+  "schemaVersion": "2.0.1",
+  "updatedAt": "2018-10-12T10:51:50.530Z"
+}
```
Comment (facuspagnuolo) [inline packages/cli/test/mocks/mock-stdlib/build/contracts/GreeterLib.json]: Is it necessary to publish these build files?

File: packages/cli/src/utils/runWithTruffle.js [hunk 2]
```diff
@@ -1,6 +1,6 @@
 import Truffle from '../models/truffle/Truffle';
 import Session from '../models/network/Session'
-import Contracts from 'zos-lib/lib/utils/Contracts';
+import { Contracts } from 'zos-lib';
```
Comment (facuspagnuolo) [inline packages/cli/src/utils/runWithTruffle.js]: 👍

File: packages/cli/src/models/status/StatusChecker.js [hunk 3]
```diff
@@ -106,23 +114,43 @@ export default class StatusChecker {
...
+    const observed = bytecodeDigest(bytecode)
+    if (observed !== expected) this.visitor.onMismatchingImplementationBodyBytecode(expected, observed, { alias, address, bodyBytecodeHash: observed })
+  }
+
+  _checkRemoteSolidityLibImplementation({ alias, address }, bytecode) {
+    if (this.networkFile.hasSolidityLib(alias)) {
+      this._checkSolidityLibImplementationAddress(alias, address)
+      this._checkSolidityLibImplementationBytecode(alias, address, bytecode)
+    }
+    //TODO: implement missing remote solidity libs validation
```
Comment (facuspagnuolo) [inline packages/cli/src/models/status/StatusChecker.js]: can we move this to an issue?

File: packages/lib/src/utils/Bytecode.js [hunk 4]
```diff
@@ -17,16 +19,38 @@ export function bytecodeDigest(rawBytecode) {
...
 }
 
+// Replaces the solidity library address inside its bytecode with zeros
+export function replaceSolidityLibAddress(bytecode, address) {
+  return bytecode.replace(address.replace(/^0x/,''), ZERO_ADDRESS.replace(/^0x/, ''))
+}
+
+// Verifies if a bytecode represents a solidity library.
+export function isSolidityLib(bytecode) {
+  return bytecode.match(/^0x73[A-Fa-f0-9]{40}3014/)
```
Comment (facuspagnuolo) [inline packages/lib/src/utils/Bytecode.js]: we should add a comment explaining why we assume that expected bytecode to be a lib

File: packages/cli/src/models/dependency/Dependency.js [hunk 5]
```diff
@@ -30,12 +30,29 @@ export default class Dependency {
...
+
+    const libraries = _.fromPairs(await Promise.all(_.map(libraryNames, async (libraryName) => {
+      const implementation = await project.setImplementation(Contracts.getFromNodeModules(this.name, libraryName), libraryName)
+      return [libraryName, implementation.address];
+    })));
+
+    await Promise.all(_.map(contracts, async ([contractClass, contractAlias]) => {
+      contractClass.link(libraries);
+      await project.setImplementation(contractClass, contractAlias);
+    }));
```
Comment (facuspagnuolo) [inline packages/cli/src/models/dependency/Dependency.js]: This code is not that clear, +1 to refactor it in a future

File: packages/lib/src/utils/Contracts.js [hunk 6]
```diff
@@ -97,6 +97,7 @@ export default {
     contract.setProvider(web3.currentProvider)
     contract.defaults({ from: web3.eth.accounts[0], ... defaults })
     contract.synchronization_timeout = syncTimeout
+    contract.setNetwork('test')
```
Comment (facuspagnuolo) [inline packages/lib/src/utils/Contracts.js]: 🙌

File: tests/cli-app/test/project.js [hunk 7]
```diff
@@ -24,12 +24,12 @@ function registerProjectHooks (network) {
...
-    run('rm contracts/*.sol ||:')
-    run('rm zos.* ||:')
-    run('rm package.* ||:')
-  })
+  // after('cleaning up project folder', function () {
+  //   run('rm build/contracts/*.json ||:')
+  //   run('rm contracts/*.sol ||:')
+  //   run('rm zos.* ||:')
+  //   run('rm package.* ||:')
+  // })
```
Comment (facuspagnuolo) [inline tests/cli-app/test/project.js]: should we remove this `after` statement?

File: packages/cli/test/mocks/mock-stdlib/build/contracts/GreeterLib.json (hunk 1 above)
Comment (spalladino) [reply packages/cli/test/mocks/mock-stdlib/build/contracts/GreeterLib.json]: Actual ~stdlibs~ ~libs~ evm packags have their build artifacts published, so these ones for testing should have them as well.

File: packages/cli/src/models/status/StatusChecker.js (hunk 3 above)
Comment (spalladino) [reply packages/cli/src/models/status/StatusChecker.js]: https://github.com/zeppelinos/zos/issues/244

File: packages/cli/src/models/dependency/Dependency.js (hunk 5 above)
Comment (spalladino) [reply packages/cli/src/models/dependency/Dependency.js]: Yup, it's on my _technical debt_ workflowy section

File: tests/cli-app/test/project.js (hunk 7 above)
Comment (spalladino) [reply tests/cli-app/test/project.js]: No, I unfortunately forgot to uncomment it. It's no harm though.

**Pull request merged**